bootloader/lkenv/lkenv.go

```diff
@@ -287,21 +336,120 @@ func (l *Env) GetBootPartition(kernel string) (string, error) {
 // returned, otherwise the reference is removed and nil is returned.
 // Note that to persist this change the env must be saved afterwards with Save.
 func (l *Env) RemoveKernelRevisionFromBootPartition(kernel string) error {
-	for x := range l.env.Bootimg_matrix {
-		if "" != cToGoString(l.env.Bootimg_matrix[x][MATRIX_ROW_PARTITION][:]) {
-			if kernel == cToGoString(l.env.Bootimg_matrix[x][MATRIX_ROW_KERNEL][:]) {
-				l.env.Bootimg_matrix[x][1][MATRIX_ROW_PARTITION] = 0
-				return nil
-			}
-		}
-	}
-	return fmt.Errorf("cannot find defined [%s] boot image partition", kernel)
+	return l.variant.removeKernelFromBootPart(kernel)
 }
 
 // GetBootImageName return expected boot image file name in kernel snap
 func (l *Env) GetBootImageName() string {
-	if "" != cToGoString(l.env.Bootimg_file_name[:]) {
-		return cToGoString(l.env.Bootimg_file_name[:])
+	fn := l.Get("bootimg_file_name")
+	if fn != "" {
+		return fn
 	}
 	return BOOTIMG_DEFAULT_NAME
 }
+
+// common matrix helper methods which take the matrix as input, then return an
+// updated version to be re-assigned to the original struct
+
+func commonConfigureBootPartitions(matr bootimgKernelMatrix, bootPartLabels []string) (bootimgKernelMatrix, error) {
+	numBootPartLabels := len(bootPartLabels)
+
+	if numBootPartLabels != SNAP_BOOTIMG_PART_NUM {
+		return matr, fmt.Errorf("invalid number of boot partition labels, expected %d got %d", SNAP_BOOTIMG_PART_NUM, numBootPartLabels)
+	}
+	copyString(matr[0][MATRIX_ROW_PARTITION][:], bootPartLabels[0])
+	copyString(matr[1][MATRIX_ROW_PARTITION][:], bootPartLabels[1])
+	return matr, nil
+}
+
+func commonRemoveKernelFromBootPart(matr bootimgKernelMatrix, kernel string) (bootimgKernelMatrix, error) {
+	for x := range matr {
+		if "" != cToGoString(matr[x][MATRIX_ROW_PARTITION][:]) {
+			if kernel == cToGoString(matr[x][MATRIX_ROW_KERNEL][:]) {
+				matr[x][1][MATRIX_ROW_PARTITION] = 0
+				return matr, nil
+			}
+		}
+	}
+
+	return matr, fmt.Errorf("cannot find kernel %q in boot image partitions", kernel)
+}
+
+func commonSetBootPartition(matr bootimgKernelMatrix, bootpart, kernel string) (bootimgKernelMatrix, error) {
+	for x := range matr {
+		if bootpart == cToGoString(matr[x][MATRIX_ROW_PARTITION][:]) {
+			copyString(matr[x][MATRIX_ROW_KERNEL][:], kernel)
+			return matr, nil
+		}
+	}
+
+	return matr, fmt.Errorf("cannot find defined [%s] boot image partition", bootpart)
+}
+
+func commonFindFreeBootPartition(env envVariant, matr bootimgKernelMatrix, kernel string) (string, error) {
+	for x := range matr {
+		bp := cToGoString(matr[x][MATRIX_ROW_PARTITION][:])
+		if bp != "" {
+			k := cToGoString(matr[x][MATRIX_ROW_KERNEL][:])
+			// return this one if it's not the current snap_kernel, if it's the
+			// exactly specified kernel, or if it's empty
+			if k != env.get("snap_kernel") || k == kernel || k == "" {
```

Comment:
env.get("kernel") seems something to be passed in instead of env, also see comment about changing these for even larger reuse